### PR TITLE
[SYCLomatic] Support migration of 14 asm instructions

### PR DESCRIPTION
**Pull request by yihwang-nv**

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 1]
```diff
@@ -556,6 +564,31 @@ struct sub_sat {
   }
 };
 
+// x << offset
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// x << offset
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 2]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
   return detail::extend_binary<RetT, true>(a, b, c, maximum(), second_op);
 }
 
+/// Extend \p a and \p b to 33 bit and return a << clamp(b, 0, 32)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend \p a and \p b to 33 bit and return a << clamp(b, 0, 32).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 3]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \returns a << clamp(b, 0, 32)
+template <typename RetT, typename T>
+inline constexpr RetT extend_shl_clamp(T a, uint32_t b) {
+  return detail::extend_binary<RetT, false>(a, sycl::clamp(b, 0u, 32u),
+                                            shift_left());
+}
+
+/// Extend Inputs to 33 bit, and return second_op(a << clamp(b, 0, 32), c)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend Inputs to 33 bit, and return second_op(a << clamp(b, 0, 32), c).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 4]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] second_op The operation to do with the third value
+/// \returns second_op(sat(a << clamp(b, 0, 32)), c)
+template <typename RetT, typename T, typename BinaryOperation>
+inline constexpr RetT extend_shl_sat_clamp(T a, uint32_t b, uint32_t c,
+                                           BinaryOperation second_op) {
+  return detail::extend_binary<RetT, true>(a, sycl::clamp(b, 0u, 32u), c,
+                                           shift_left(), second_op);
+}
+
+/// Extend \p a and \p b to 33 bit and return a << (b & 0x1F)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend \p a and \p b to 33 bit and return a << (b & 0x1F).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 5]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \returns a << (b & 0x1F)
+template <typename RetT, typename T>
+inline constexpr RetT extend_shl_warp(T a, uint32_t b) {
+  return detail::extend_binary<RetT, false>(a, b & 0x1F, shift_left());
+}
+
+/// Extend Inputs to 33 bit, and return second_op(a << (b & 0x1F), c)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend Inputs to 33 bit, and return second_op(a << (b & 0x1F), c).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 6]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] second_op The operation to do with the third value
+/// \returns second_op(sat(a << (b & 0x1F)), c)
+template <typename RetT, typename T, typename BinaryOperation>
+inline constexpr RetT extend_shl_sat_warp(T a, uint32_t b, uint32_t c,
+                                          BinaryOperation second_op) {
+  return detail::extend_binary<RetT, true>(a, b & 0x1F, c, shift_left(),
+                                           second_op);
+}
+
+/// Extend \p a and \p b to 33 bit and return a >> clamp(b, 0, 32)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend \p a and \p b to 33 bit and return a >> clamp(b, 0, 32).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 7]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \returns a >> clamp(b, 0, 32)
+template <typename RetT, typename T>
+inline constexpr RetT extend_shr_clamp(T a, uint32_t b) {
+  return detail::extend_binary<RetT, false>(a, sycl::clamp(b, 0u, 32u),
+                                            shift_right());
+}
+
+/// Extend Inputs to 33 bit, and return second_op(a >> clamp(b, 0, 32), c)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend Inputs to 33 bit, and return second_op(a >> clamp(b, 0, 32), c).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 8]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] second_op The operation to do with the third value
+/// \returns second_op(sat(a >> clamp(b, 0, 32)), c)
+template <typename RetT, typename T, typename BinaryOperation>
+inline constexpr RetT extend_shr_sat_clamp(T a, uint32_t b, uint32_t c,
+                                           BinaryOperation second_op) {
+  return detail::extend_binary<RetT, true>(a, sycl::clamp(b, 0u, 32u), c,
+                                           shift_right(), second_op);
+}
+
+/// Extend \p a and \p b to 33 bit and return a >> (b & 0x1F)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend \p a and \p b to 33 bit and return a >> (b & 0x1F).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 9]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \returns a << (b & 0x1F)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \returns a >> (b & 0x1F)
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 10]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+template <typename RetT, typename T>
+inline constexpr RetT extend_shr_warp(T a, uint32_t b) {
+  return detail::extend_binary<RetT, false>(a, b & 0x1F, shift_right());
+}
+
+/// Extend Inputs to 33 bit, and return second_op(a >> (b & 0x1F), c)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// Extend Inputs to 33 bit, and return second_op(a >> (b & 0x1F), c).
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 11]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \param [in] c The value to merge
+/// \param [in] second_op The operation to do with the third value
+/// \returns second_op(a << (b & 0x1F), c)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \returns second_op(a >> (b & 0x1F), c)
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 12]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+inline constexpr RetT extend_shr_warp(T a, uint32_t b, uint32_t c,
+                                      BinaryOperation second_op) {
+  return detail::extend_binary<RetT, false>(a, b & 0x1F, c, shift_right(),
+                                            second_op);
+}
+
+/// Extend \p a and \p b to 33 bit and return sat(a >> (b & 0x1F)).
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \returns sat(a << (b & 0x1F))
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \returns sat(a >> (b & 0x1F))
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 13]
```diff
@@ -1232,6 +1265,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+inline constexpr RetT extend_shr_sat_warp(T a, uint32_t b) {
+  return detail::extend_binary<RetT, true>(a, b & 0x1F, shift_right());
+}
+
+/// Extend Inputs to 33 bit, and return second_op(sat(a >> (b & 0x1F)), c).
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \param [in] c The value to merge
+/// \param [in] second_op The operation to do with the third value
+/// \returns second_op(sat(a << (b & 0x1F)), c)
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \returns second_op(sat(a >> (b & 0x1F)), c)
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 14]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
   return detail::extend_vbinary2<RetT, true, false>(a, b, c, maximum());
 }
 
+/// Compute vectorized average of \p a and \p b, with each value treated as a 2
+/// elements vector type and extend each element to 17 bit.
+/// \tparam [in] RetT The type of the return value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] RetT The type of the return value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 15]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \tparam [in] AT The type of the first value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] AT The type of the first value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 16]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \tparam [in] BT The type of the second value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] BT The type of the second value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 17]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \returns The extend vectorized average of the two values
+template <typename RetT, typename AT, typename BT>
+inline constexpr RetT extend_vavrg2(AT a, BT b, RetT c) {
+  return detail::extend_vbinary2<RetT, false, false>(a, b, c, average());
+}
+
+/// Compute vectorized average of \p a and \p b, with each value treated as a 2
+/// elements vector type and extend each element to 17 bit. Then add each half
+/// of the result and add with \p c.
+/// \tparam [in] RetT The type of the return value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] RetT The type of the return value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 18]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \tparam [in] AT The type of the first value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] AT The type of the first value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 19]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \tparam [in] BT The type of the second value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] BT The type of the second value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 20]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \returns The addition of each half of extend average maximum of the
+/// two values and the third value
+template <typename RetT, typename AT, typename BT>
+inline constexpr RetT extend_vavrg2_add(AT a, BT b, RetT c) {
+  return detail::extend_vbinary2<RetT, false, true>(a, b, c, average());
+}
+
+/// Compute vectorized average of \p a and \p b with saturation, with each value
+/// treated as a 2 elements vector type and extend each element to 17 bit.
+/// \tparam [in] RetT The type of the return value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] RetT The type of the return value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 21]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \tparam [in] AT The type of the first value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] AT The type of the first value, can only be 32 bit integer
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 22]
```diff
@@ -1452,6 +1665,50 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \tparam [in] BT The type of the second value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// \tparam [in] BT The type of the second value, can only be 32 bit integer
```

**Review by tangjj11 — CHANGES_REQUESTED** [review]

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 22 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 21 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 20 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 19 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 18 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 17 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 16 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 15 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 14 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 13 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 12 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 11 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 10 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 9 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 8 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 7 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 6 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 5 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 4 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 3 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 2 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 1 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 17 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: have not fixed.

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 18 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: have not fixed.

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 19 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: have not fixed.

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 20 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: have not fixed.

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 21 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: have not fixed.

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 22 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: have not fixed.

**Review by tangjj11 — CHANGES_REQUESTED** [review]

**Review by tangjj11 — APPROVED** [review]

File: clang/test/dpct/asm/copysign.cu [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+
+// clang-format off
+// CHECK: #include <cmath>
+#include <cuda_runtime.h>
+#include <cstdint>
+
+__global__ void copysign() {
+  float f32;
+  double f64;
+  // CHECK: f32 = std::copysign(10.0f, -100.0f);
```
Comment (zhimingwang36) [inline clang/test/dpct/asm/copysign.cu]: let's use sycl::copysign?

File: clang/test/dpct/asm/ex2.cu [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+// RUN: FileCheck %s --match-full-lines --input-file %T/ex2/ex2.dp.cpp
+// RUN: %if build_lit %{icpx -c -fsycl %T/ex2/ex2.dp.cpp -o %T/ex2/ex2.dp.o %}
+
+// clang-format off
+#include <cuda_runtime.h>
+#include <cstdint>
+
+__global__ void ex2() {
+  float f32;
+  // CHECK: f32 = sycl::pow<float>(2, 1.0f);
```
Comment (zhimingwang36) [inline clang/test/dpct/asm/ex2.cu]: for constant 1.0 value, let' sycl::pown(2, 1)

File: clang/test/dpct/asm/selp.cu [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+  int64_t i64;
+  uint64_t u64;
+  float f32;
+  double f64;
+
+  // CHECK: i16 = i16 == 1 ? i16 : i16;
+  asm("selp.s16 %0, %1, %2, %3;" : "=h"(i16) : "h"(i16), "h"(i16), "h"(i16));
+  
+  // CHECK: u16 = u16 == 1 ? u16 : u16;
+  asm("selp.u16 %0, %1, %2, %3;" : "=h"(u16) : "h"(u16), "h"(u16), "h"(u16));
```
Comment (zhimingwang36) [inline clang/test/dpct/asm/selp.cu]: better to update the arg names to make sure the arg list in migrated code is correct.

File: clang/test/dpct/asm/sqrt.cu [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+__global__ void sqrt() {
+  float f32;
+  double f64;
+  // CHECK: f32 = sycl::sqrt<float>(1.0f);
+  asm("sqrt.approx.f32 %0, %1;" : "=f"(f32) : "f"(1.0f));
+
+  // CHECK: f32 = sycl::sqrt<float>(1.0f);
+  asm("sqrt.approx.ftz.f32 %0, %1;" : "=f"(f32) : "f"(1.0f));
+  
+  // CHECK: f32 = sycl::vec<float, 1>(sycl::sqrt<float>(1.0f)).convert<float, sycl::rounding_mode::rte>()[0];
```
Comment (zhimingwang36) [inline clang/test/dpct/asm/sqrt.cu]: sycl::sqrt<float>(1.0f))  , and emit a message ignore round mode

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 27]
```diff
@@ -1232,6 +1255,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+  return detail::extend_binary<RetT, true>(a, sycl::clamp(b, 0u, 32u), c,
+                                           detail::shift_left(), second_op);
+}
+
+/// Extend \p a and \p b to 33 bit and return a << (b & 0x1F).
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \returns a << (b & 0x1F)
+template <typename RetT, typename T>
+inline constexpr RetT extend_shl_warp(T a, uint32_t b) {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
inline constexpr RetT extend_shl_mask31(T a, uint32_t b) {
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 28]
```diff
@@ -1232,6 +1255,186 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+}
+
+/// Extend Inputs to 33 bit, and return second_op(a << (b & 0x1F), c).
+/// \param [in] a The source value
+/// \param [in] b The offset to shift
+/// \param [in] c The value to merge
+/// \param [in] second_op The operation to do with the third value
+/// \returns second_op(a << (b & 0x1F), c)
+template <typename RetT, typename T, typename BinaryOperation>
+inline constexpr RetT extend_shl_warp(T a, uint32_t b, uint32_t c,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: same as above.

File: clang/test/dpct/asm/copysign.cu (hunk 23 above)
Comment (yihwang-nv) [reply clang/test/dpct/asm/copysign.cu]: done

File: clang/test/dpct/asm/ex2.cu (hunk 24 above)
Comment (yihwang-nv) [reply clang/test/dpct/asm/ex2.cu]: done

File: clang/test/dpct/asm/selp.cu (hunk 25 above)
Comment (yihwang-nv) [reply clang/test/dpct/asm/selp.cu]: done

File: clang/test/dpct/asm/sqrt.cu (hunk 26 above)
Comment (yihwang-nv) [reply clang/test/dpct/asm/sqrt.cu]: done

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 28 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: done

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**